escrow/state_service.go

```diff
@@ -51,6 +51,29 @@ func (service *PaymentChannelStateService) GetChannelState(context context.Conte
 		return nil, errors.New("only channel signer can get latest channel state")
 	}
 
+	// check if nonce matches with blockchain or not
+	nonceEqual, err := service.channelService.StorageNonceMatchesWithBlockchainNonce(&PaymentChannelKey{ID: channelID})
+	if err != nil {
+		log.WithError(err).Infof("payment data not available in payment storage.")
+	} else if !nonceEqual {
+		// check for payments in the payment storage with current nonce -1, this will happen  cli has issues in claiming process
+		paymentID := fmt.Sprintf("%v/%v", channel.ChannelID, (&big.Int{}).Sub(channel.Nonce, big.NewInt(1)))
+		payment, ok, err := service.paymentStorage.Get(paymentID)
+		if err == nil && ok && payment != nil {
+			log.Infof("old payment detected for the channel %v (payments with  nonce = current nonce -1).", channelID)
+
+			// return the channel state with old nonce
+			return &ChannelStateReply{
+				CurrentNonce:         bigIntToBytes(channel.Nonce),
+				CurrentSignedAmount:  bigIntToBytes(channel.AuthorizedAmount),
+				CurrentSignature:     channel.Signature,
+				OldNonceSignedAmount: bigIntToBytes(payment.Amount),
+				OldNonceSignature:    payment.Signature,
+			}, nil
+		}
+		log.WithError(err).Infof("unable to extract old payment from storage" )
```

Comment:
I would separately detect and handle three cases here:
```go
  if err != nil {
    log.WithError(err).Errorf("unable to extract old payment from storage" )
  }
  if !ok {
    if /* blockchain nonce is not equal to channel nonce, result can be different because of data race */ {
      log.Errorf("old payment is not found in storage, nevertheless local channel nonce is not equal to the blockchain one, channel: %v" channelID)
      return nil, errors.New("channel has different nonce in local storage and blockchain")
    }
  }
  return &ChannelStateReply{
    CurrentNonce:         bigIntToBytes(channel.Nonce),
    CurrentSignedAmount:  bigIntToBytes(channel.AuthorizedAmount),
    CurrentSignature:     channel.Signature,
    OldNonceSignedAmount: bigIntToBytes(payment.Amount),
    OldNonceSignature:    payment.Signature,
  }, nil
```

Reply:
fixed this as well